File: src/doubleDispatch/java/SpaceSimulatorClient.java
package doubleDispatch.java;

public class SpaceSimulatorClient {

    public static void main(String[] args){
        ImperialShip unscannedShip = new ImperialShip();
        ImperialShip enemyUnscannedShip = new ImperialShip();

        ImperialShip scannedShip = new DauntlessCruiser();
        ImperialShip enemyScannedShip = new DauntlessCruiser();

        DauntlessCruiser enemyShipInVisualRange = new DauntlessCruiser();

        unscannedShip.fireUpon(enemyUnscannedShip);
        unscannedShip.fireUpon(enemyScannedShip);
        unscannedShip.fireUpon(enemyShipInVisualRange);

        scannedShip.fireUpon(enemyUnscannedShip);
        scannedShip.fireUpon(enemyScannedShip);
        scannedShip.fireUpon(enemyShipInVisualRange);

    }
}

File: src/doubleDispatch/csharp/SpaceSimulatorClient.cs
namespace doubleDispatch.csharp{
    class SpaceSimulatorClient {

        public static void Main(string[] args) {
            OrkShip unscannedShip = new OrkShip();
            OrkShip enemyUnscannedShip = new OrkShip();

            OrkShip scannedShip = new OrkSavageGunship();
            OrkShip enemyScannedShip = new OrkSavageGunship();

            OrkSavageGunship enemyShipInVisualRange = new OrkSavageGunship();

            unscannedShip.fireUpon((dynamic) enemyUnscannedShip);
            unscannedShip.fireUpon((dynamic) enemyScannedShip);
            unscannedShip.fireUpon((dynamic) enemyShipInVisualRange);

            ((dynamic) scannedShip).fireUpon((dynamic) enemyUnscannedShip);
            ((dynamic) scannedShip).fireUpon((dynamic) enemyScannedShip);
            ((dynamic) scannedShip).fireUpon((dynamic) enemyShipInVisualRange);

        }
    }
}
